pontoon/terminology/models.py

```diff
@@ -0,0 +1,251 @@
+            term_words = term_singulars.split()
+            term_len = len(term_words)
+
+            while True:
+                try:
+                    scan_start += (entity_singular_words[scan_start:]).index(term_words[0])
+                    scan_end = scan_start + term_len
+                    if entity_singular_words[scan_start:scan_end] == term_words:
+                        phrase = ' '.join(
+                            original_string_words[scan_start:scan_end]
+                        )
+                        terms_phrases[phrase].update(term_pks)
+                        del entity_singular_words[scan_start:scan_end]
+                        del original_string_words[scan_start:scan_end]
+                    else:
+                        scan_start = scan_end
+                except ValueError:
+                    break
+
+        return terms_phrases.items()
+
+    @transaction.atomic
+    def assign_terms_to_entities(self, entities):
+        entity_terms = []
+        entities_pks = [e[0] for e in entities]
+        # Remove existing terms relations.
+        EntityTerm.objects.filter(entity__in=entities_pks).delete()
+        for entity_pk, string, string_singulars, string_plural, string_plural_singulars in entities:
+            if string_singulars:
+                entity_terms.extend([
+                    (entity_pk, term_pks, phrase)
+                    for (phrase, term_pks) in self.find_terms(string, string_singulars)
+                ])
+
+            if string_plural_singulars:
+                entity_terms.extend([
+                    (entity_pk, term_pks, phrase)
+                    for (phrase, term_pks) in self.find_terms(
+                        string_plural, string_plural_singulars)
+                ])
+        terms_db = []
+
+        for entity_pk, terms, phrase in entity_terms:
+            for term_pk in terms:
+                terms_db.append(
+                    EntityTerm(entity_id=entity_pk, term_id=term_pk, phrase=phrase)
+                )
+        EntityTerm.objects.bulk_create(terms_db, 10000)
+
+        terms_count = (
+            Term.objects.filter(entityterm__entity__pk__in=entities_pks)
+            .distinct()
+            .count()
+        )
+
+        phrases_count = (
+            EntityTerm.objects.filter(entity__pk__in=entities_pks)
+            .distinct()
+            .count()
+        )
+
+        log.info(
+            'Assigned %s terms to %s entities and matched %s phrases.',
+            terms_count,
+            len(entities),
+            phrases_count
+        )
+
+    @transaction.atomic
+    def import_tbx_terms(self, tbx_terms):
+        """
+        Creates/updates term objects passed from terminology files.
+        All terms will be automatically matched with their entities.
+
+        :param VCSTerm tbx_terms: a list of terms imported from the filesystem/vcs.
+        """
+        db_terms = {}
+
+        for term in Term.objects.all():
```

Comment:
This is kinda worrisome. How big can this table become? Could this lead to out-of-memory problems? Are we sure it cannot?